HEC-1426 CTUTR attempts blocking functionality

meghana-viswanath opened the pull request:
Details in ticket -> https://jira.tools.tax.service.gov.uk/browse/HEC-1426

priyankadhiman3006 commented on app/uk/gov/hmrc/hecapplicantfrontend/controllers/CompanyDetailsController.scala:
```diff
@@ -307,74 +308,81 @@ class CompanyDetailsController @Inject() (
+              .copy(ctutr = Some(ctutr))
+            val (start, end)   = CompanyDetailsController.calculateLookBackPeriod(timeProvider.currentDate)
+
+            val ctutrAttemptsDeleteFut = ctutrAttemptsService.delete(crn, companySession.loginData.ggCredId)
```
it could have been included directly into the for comprehension below, so as ctStatusFut

meghana-viswanath replied:
my reasoning [here](https://github.com/hmrc/hec-applicant-frontend/pull/85#discussion_r751096891), we can continue the discussion there?

AndyHWChung commented on app/uk/gov/hmrc/hecapplicantfrontend/controllers/CompanyDetailsController.scala:
```diff
@@ -307,74 +308,81 @@ class CompanyDetailsController @Inject() (
-            val updatedSession = companySession
-              .copy(ctutrAnswerAttempts = companySession.ctutrAnswerAttempts + 1, userAnswers = updatedAnswers)
+              val result = for {
+                ctutrAttempts <- ctutrAttemptsService.createOrIncrementAttempts(crn, companySession.loginData.ggCredId)
```
if the maximum number of attempts has already been reached then I think with the current set up the `lastUpdated` field in `CtutrAttempts` will be updated. We'll eventually need the time where the lock expires, so maybe instead of
```scala
final case class CtutrAttempts(
  crn: CRN,
  ggCredId: GGCredId,
  attempts: Int,
  lastUpdated: ZonedDateTime
)
```
we could have something like
```scala
final case class CtutrAttempts(
  crn: CRN,
  ggCredId: GGCredId,
  attempts: Int,
  lockedUntil: Option[ZonedDateTime]
)
```
Where `lockedUntil` is populated as soon as the maximum number of attempts has been reached and not updated any time afterwards. That means we'd be able to tell if someone has been locked out by looking at the presence of that field instead of looking at the number of attempts - this means it'd be more resilient to changes of the maximum value if we ever change it - people who are locked out under an old value will still be locked out if we increase the value. Also, when we come to display the time they can try again, we don't need to do a calculation based upon the last updated time.

meghana-viswanath replied:
yeah, I see your point, we can get hold of `lastUpdated` anyway from the context data, so makes sense to pre-calculate and store 👍

meghana-viswanath replied:
updated https://github.com/hmrc/hec-applicant-frontend/pull/85/commits/5fc7cc1a19ceda4df7804ce3510a700a4b57dcde

meghana-viswanath commented on app/uk/gov/hmrc/hecapplicantfrontend/controllers/CompanyDetailsController.scala:
```diff
@@ -307,74 +308,81 @@ class CompanyDetailsController @Inject() (
+            val (start, end)   = CompanyDetailsController.calculateLookBackPeriod(timeProvider.currentDate)
+
+            val ctutrAttemptsDeleteFut = ctutrAttemptsService.delete(crn, companySession.loginData.ggCredId)
+            val ctStatusFut            = taxCheckService.getCTStatus(ctutr.strippedCtutr, start, end)
```
My thinking was to kick both the futures off simultaneously and save time, doing them directly in the for-comprehension will mean waiting for the first future before doing the second future. Agree that in the case that one of them fails, the other will be done, but those are exceptional cases and won't happen as often, so leaned towards saving time here

AndyHWChung replied:
ah ok - I get your logic here. In theory this shouldn't cause a lot of issues cos the call to delete failing is going to be rare, so a wasted call to the CT status API is rare. I think I'm ok leaving this as is then and seeing if it causes any issues in the future (which I don't think will be the case). What do you think @priyankadhiman3006?

priyankadhiman3006 replied:
yup, this will not cause a issue  , just the call to fetch ctStats will get wasted , even if the delete fails. no issues in keeping the existing code

priyankadhiman3006 replied:
I was wondering, when should we delete the attempts from the DB, should it be done in the beginning and then session is updated as done here , or should be done at the end , so even if the db delete is failed, we have correct value in the session?

meghana-viswanath replied:
not sure I understand, as it is now, if the delete fails, a `RuntimeException` is thrown and the session is left as is

priyankadhiman3006 replied:
nvm, pls ignore

priyankadhiman3006 commented on app/uk/gov/hmrc/hecapplicantfrontend/services/CtutrAttemptsService.scala:
```diff
@@ -0,0 +1,68 @@
+  /**
+    * Fetch CTUTR attempts for the CRN-GGCredId combination
+    * If found, increment the number of attempts
+    * If not found, create a new document with 0 attempts
```
do we need to create a new doc with 0 attempts? cause as per the discussion during refinement when we are checking the db and  there is no data it means it is not blocking. I may be wrong , but just want to double check

meghana-viswanath replied:
no, you are right, the docstring is incorrect, it creates a new doc and sets the number of attempts to 1, I'll fix the comment 👍

meghana-viswanath replied:
updated https://github.com/hmrc/hec-applicant-frontend/pull/85/commits/5fc7cc1a19ceda4df7804ce3510a700a4b57dcde

AndyHWChung commented on app/uk/gov/hmrc/hecapplicantfrontend/services/CtutrAttemptsService.scala:
```diff
@@ -0,0 +1,76 @@
+      updatedAttempts = attemptsOpt match {
+                          case Some(ctutrAttempts) =>
+                            val lockedUntilOpt = if (ctutrAttempts.attempts >= appConfig.maxCtutrAnswerAttempts - 1) {
+                              val durationInSeconds = appConfig.maxCtutrStoreExpiry.toSeconds
```
`appConfig.maxCtutrStoreExpiry.toSeconds` could be taken out of this `def` so we don't need to do the calculation to seconds every time

AndyHWChung commented on test/uk/gov/hmrc/hecapplicantfrontend/utils/Fixtures.scala:
```diff
@@ -205,4 +209,36 @@ object Fixtures {
     ctStatus = ctStatus
   )
 
+  class TestConfig(
```
this seems like quite a big thing to maintain - we could maybe explore using scala mock to create a mock instance of AppConfig and configure it to return certain values of val's like we do in the controller specs

meghana-viswanath replied:
Yeah, we would have to add a new field/method in 3 places as opposed to one. In controller specs we extend `ControllerSpec` and override the `additionalConfig` like this - 
```
override lazy val additionalConfig: Configuration = Configuration(
    ConfigFactory.parseString(s"ctutr-attempts.maximum-attempts = $maxCtutrAttempts")
  )
```
But here, we want to inject `AppConfig` which is not the same.

The problem is that `scalamock` does not allow mocking `val`s - their workaround suggestion is to use `def`s instead and I didn't want to do that. If we did use `def`s, we can do the below and get rid of 👆 
```
val mockAppConfig = mock[AppConfig]
(mockAppConfig.maxCtutrAnswerAttempts _).expects().returning(3)
(mockAppConfig.maxCtutrStoreExpiry _).expects().returning(3 hours)
```

Gladly welcome any ideas here because I'm not super happy with this either.

AndyHWChung replied:
I think my suggestion in my other comment might help where we may be able to circumvent the need for injecting in `AppConfig` entirely

AndyHWChung commented on app/uk/gov/hmrc/hecapplicantfrontend/controllers/CompanyDetailsController.scala:
```diff
@@ -307,74 +308,81 @@ class CompanyDetailsController @Inject() (
-          )
+        ensureUserAnswersHasCRN(companySession) { crn =>
+          def maxCtutrAnswerAttemptsReached(companySession: CompanyHECSession) =
+            companySession.ctutrAnswerAttempts >= appConfig.maxCtutrAnswerAttempts
```
this check doesn't work for different CRN's now - I don't think we need `ctutrAnswerAttempts` in session anymore. Here, we could do a get on the `CtutrAttemptsStore` with the `CRN` and `ggCredId` and check if `lockedUntil` exists to see if the maximum number of attempts has been reached for this particular CRN. If we do that, then this controller doesn't need to know the value of `appConfig.maxCtutrAnswerAttempts` and then we can remove the CT UTR attempts config values from `AppConfig` and just inject `Configuration` into `CtutrAttemptsService`. Then we can leave `AppConfig` as it was before

meghana-viswanath replied:
think you are right, I'll give a go, I think we'll still need something in the session to indicate the presence of a lock so the `JourneyService` can determine what to do next

meghana-viswanath commented:
@AndyHWChung @priyankadhiman3006  I've made the following changes based on comments from yesterday
- replaced `CtutrAttempts.lastUpdated` with `blockedUntil` that gets populated only when the max allowed attempts is reached. Once this field is populated, the value in ctutr attempts field is irrelevant.
- instead of injecting `AppConfig` into `CtutrAttemptsServiceImpl`, I've taken Andy's suggestion and injected `Configuration` instead allowing us to revert to the original simpler state of `AppConfig`
- replaced `CompanyHECSession.ctutrAttempts` with `crnBlocked` boolean field. We need a way to understand when a CRN is blocked within the `JourneyService` so this information needs to be in the session. But a simple boolean indicating whether or not the CRN in the answers is blocked should be sufficient.

Let me know your thoughts.

AndyHWChung commented on app/uk/gov/hmrc/hecapplicantfrontend/controllers/CompanyDetailsController.scala:
```diff
@@ -307,99 +308,88 @@ class CompanyDetailsController @Inject() (
+                    case ctutrAttempts if ctutrAttempts.isBlocked =>
+                      updateAndNextJourneyData(
+                        routes.CompanyDetailsController.enterCtutr(),
+                        companySession.copy(crnBlocked = ctutrAttempts.isBlocked)
```
`isBlocked` and the `crnBlocked` needs to know which CRN is being referred to - you can in theory block yourself on one CRN but still have three goes on another CRN. 

That may be that `crnBlocked` becomes something like `blockedCRNs: List[CRN]`. At some point it seems like 
the thing in session needs to be cleared since it'll not be blocked forever. So maybe it needs to be more something like `blockedCRNs: List[(CRN, ZonedDateTime)]` where it some case class instead of a tuple and the time refers to the time the block is lifted. Then you can know for certain when you're in the `JourneyService` that the lock hasn't expired.

meghana-viswanath replied:
`isBlocked` is against `CRN` and `GGCredId` combination, so we should be ok there I think.

The `crnBlocked` field on the session will be against the `CRN` in the user answers. When the answer changes, the `crnBlocked` also changes with it by fetching it from mongo and setting it on the session so it can be used in the `JourneyService`. I'm not certain that we need a list in the session.

AndyHWChung replied:
ah ok, yeah I see with `isBlocked` now.

With `crnBlocked` wouldn't we need some changes in `CRNControlelr` on the submit endpoint then to make it reset so we know for sure that `crnBlocked` is for the CRN in session?

meghana-viswanath replied:
There's another ticket [HEC-1427](https://jira.tools.tax.service.gov.uk/browse/HEC-1427) which has enhancements to the `CRNController` where we do a check to see if the CRN is blocked and proceed based on that. For now, if CRN answer changes, the `crnBlocked` is reset to `false` and user is allowed to carry on

AndyHWChung replied:
to reset it to `false` wouldn't you need something like `session.copy(crnBlocked = false)` in the `CRNController` submit endpoint [here](https://github.com/hmrc/hec-applicant-frontend/blob/mv-hec-1426-blocking/app/uk/gov/hmrc/hecapplicantfrontend/controllers/CRNController.scala#L96)

meghana-viswanath replied:
yeah, that's right, it doesn't get reset, I got confused, but when we've implemented the HEC-1427, we'll do it then, does that sound ok?

so when the CRN answer changes, in the `CRNController`, we check if there's a block against it and set the `crnBlocked` values based on that

meghana-viswanath replied:
would it make sense to do that ticket along with this work?

AndyHWChung replied:
We could do it in either I guess, it should get done eventually. In 1427, in scenario 1 there's mention of being able to proceed if the CRN is not blocked, but there's no mention of being able to do a number of attempts on the CT UTR - perhaps that should be added in that scenario to make sure that that gets tested

meghana-viswanath replied:
I'll do a separate PR if that's ok @AndyHWChung

priyankadhiman3006 commented on app/uk/gov/hmrc/hecapplicantfrontend/controllers/CompanyDetailsController.scala:
```diff
@@ -307,99 +308,88 @@ class CompanyDetailsController @Inject() (
 
-        if (maxCtutrAnswerAttemptsReached(companySession)) goToNextPage(companySession)
-        else {
           enterCtutrForm(desCtutr)
```
There should be a check before form submit, if there is a combination of crn + ggcrcedid in db , the form should not look for handling errors or check for valid answer, it should directly go to too many attempt page

meghana-viswanath replied:
ok, I see, makes sense, I'll fix 👍

meghana-viswanath replied:
updated https://github.com/hmrc/hec-applicant-frontend/pull/85/commits/cef49808c78deff4bf0a52ebe269d495f069d189

priyankadhiman3006 commented:
I think there is a bug in there, go through company journey, enter ctutr wrong two times,
enter it correct 3rd time. go to CYA page. click on change link against ctutr. it still allow you to enter wrong ctutr three more times.
As discussed earlier, We are planning to remove that change link against ctutr  from CYA page, @AndyHWChung should we consider this as bug or ignore it

AndyHWChung commented on app/uk/gov/hmrc/hecapplicantfrontend/services/CtutrAttemptsService.scala:
```diff
@@ -0,0 +1,105 @@
+      case Some(ctutrAttempts) if ctutrAttempts.isBlocked => ctutrAttempts
+      case Some(ctutrAttempts)                            =>
+        val lockedUntilOpt = if (ctutrAttempts.attempts >= maxCtutrAnswerAttempts - 1) {
+          Some(timeProvider.now.plusSeconds(maxCtutrStoreExpiryInSeconds))
```
if a CRN has already been locked, it seems to make more sense to leave the lock time alone until it expires - with this it seems to extend the lock time each time this method gets called. Perhaps this should be:
```scala
ctutrAttempts.blockedUntil
  .OrElse(Some(timeProvider.now.plusSeconds(maxCtutrStoreExpiryInSeconds)))
```

meghana-viswanath replied:
if CRN has already been locked, the first case statement is matched where it does nothing and returns as is

AndyHWChung replied:
oh yeah - sorry! Ignore me

AndyHWChung commented on app/uk/gov/hmrc/hecapplicantfrontend/services/CtutrAttemptsService.scala:
```diff
@@ -0,0 +1,105 @@
+    ggCredId: GGCredId,
+    ctutrAttempts: Option[CtutrAttempts]
+  ): EitherT[Future, models.Error, CtutrAttempts] = {
+    require(
```
this is a bit ugly having to do it this way - I wonder if there's a way to make the types encode the infromation better. What do you think about `get` not returning `Option[CtutrAttempts]` but returning just `CtutrAttempts` instead with `attempts` equal to `0` if it doesn't exist? Then we could just pass in `CtutrAttempts` attempts to this function without having to worry about mismatching CRN's and GG cred id's

meghana-viswanath replied:
I like it!

meghana-viswanath replied:
updated https://github.com/hmrc/hec-applicant-frontend/pull/85/commits/9a2cf72f8532dd7b9270caebb01a4f4b053b1845

AndyHWChung replied:
:+1:  looks neater now!

meghana-viswanath replied:
Awesome! Thanks for all the suggestions @AndyHWChung @priyankadhiman3006 !
